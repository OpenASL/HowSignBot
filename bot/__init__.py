__version__ = "21.12.1"
