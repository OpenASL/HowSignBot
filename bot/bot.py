import asyncio
import itertools
import logging

import disnake
from disnake import ApplicationCommandInteraction
from disnake.ext import commands

from . import settings

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX


intents = disnake.Intents.default()
intents.typing = False
intents.presences = False
intents.bans = False
intents.integrations = False
intents.webhooks = False
intents.invites = False
intents.members = True
intents.messages = True
bot = commands.Bot(
    command_prefix=COMMAND_PREFIX,
    case_insensitive=True,
    owner_id=settings.OWNER_ID,
    intents=intents,
    sync_commands_debug=settings.DEBUG,
    reload=settings.DEBUG,
    test_guilds=settings.TEST_GUILDS or None,
)

PRESENCE_COMMANDS = (
    "/sign",
    "/handshape",
    "/catchphrase | /codenames",
    "/schedule new",
    "/watch2gether",
    "/sentence | /idiom",
    "/feedback",
)
CHANGE_PRESENCE_EVERY = 120  # seconds


async def cycle_presence():
    for command in itertools.cycle(PRESENCE_COMMANDS):
        activity = disnake.Activity(
            name=command,
            type=disnake.ActivityType.watching,
        )
        logger.debug(f"changing presence to show command: {command}")
        await bot.change_presence(activity=activity)
        await asyncio.sleep(CHANGE_PRESENCE_EVERY)


@bot.event
async def on_ready():
    bot.loop.create_task(cycle_presence())


@bot.event
async def on_command_error(ctx, error):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await ctx.send(error)
    else:
        logger.error(f"unhandled exception from command: {ctx.command}", exc_info=error)


@bot.event
async def on_slash_command_error(inter: ApplicationCommandInteraction, error: Exception):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await inter.send(error)
    else:
        logger.error(
            f"unhandled exception from command: {inter.application_command.name!r}",
            exc_info=error,
        )


async def set_default_presence():
    activity = disnake.Activity(
        name="/sign",
        type=disnake.ActivityType.watching,
    )
    await bot.change_presence(activity=activity)
