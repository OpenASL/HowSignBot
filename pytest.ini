[pytest]
filterwarnings =
    ignore:.*@coroutine.*:DeprecationWarning
