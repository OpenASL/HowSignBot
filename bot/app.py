import asyncio
import logging

import aiohttp_cors
from aiohttp import web
from ariadne import graphql
from ariadne.constants import PLAYGROUND_HTML

from . import settings
from .bot import bot
from .database import store
from .graphql.schema import schema
from .utils.extensions import walk_extensions

logger = logging.getLogger(__name__)

# Assign app to bot so that extensions can add routes
bot.app = app = web.Application()  # type: ignore

app.cors = cors = aiohttp_cors.setup(
    app,
    defaults={
        "*": aiohttp_cors.ResourceOptions(
            allow_credentials=False,
            expose_headers="*",
            allow_headers="*",
        )
    },
)


async def ping(_):
    return web.Response(body="", status=200)


def graphql_playground(_):
    return web.Response(text=PLAYGROUND_HTML, status=200, content_type="text/html")


async def graphql_server(request):
    data = await request.json()

    success, result = await graphql(
        schema, data, context_value=request, debug=settings.DEBUG
    )

    status = 200 if success else 400
    return web.json_response(result, status=status)


app.add_routes([web.get("/ping", ping)])
resource = bot.app.router.add_resource("/graphql")  # type: ignore
cors.add(resource.add_route("POST", graphql_server))


async def start_bot():
    try:
        await bot.start(settings.DISCORD_TOKEN)
    finally:
        await bot.close()


async def on_startup(app):
    for ext in walk_extensions():
        bot.load_extension(ext)
    await store.connect()
    app["bot_task"] = asyncio.create_task(start_bot())
    app["bot"] = bot


async def on_shutdown(app):
    app["bot_task"].cancel()
    await app["bot_task"]
    await store.disconnect()


app.on_startup.append(on_startup)
app.on_shutdown.append(on_shutdown)
