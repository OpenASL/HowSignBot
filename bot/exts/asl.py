import logging
from urllib.parse import quote_plus

import discord
from discord.ext.commands import Cog, Bot, Context, command, errors

import handshapes
from bot import settings
from bot.utils import get_spoiler_text, did_you_mean

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX

SIGN_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for {word_uppercased}]({signingsavvy})
[🌐 **Spread The Sign** - {word_uppercased}]({spread_the_sign})
[📹 **YouGlish** - ASL videos with {word_uppercased}]({youglish})
Share: {howsign}
"""

SIGN_SPOILER_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for ||{word_uppercased}||]({signingsavvy})
[🌐 **Spread The Sign** - ||{word_uppercased}||]({spread_the_sign})
[📹 **YouGlish** - ASL videos with ||{word_uppercased}||]({youglish})
Share: ||{howsign}||
"""

SIGN_HELP = """Look up a word or phrase

If the word or phrase is sent in spoiler text, i.e. enclosed in `||`, the word will also be blacked out in the reply.
To search multiple words/phrases, separate the values with a comma.

Examples:
{COMMAND_PREFIX}sign tiger
{COMMAND_PREFIX}sign ||tiger||
{COMMAND_PREFIX}sign what's up
{COMMAND_PREFIX}sign church, chocolate, computer
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def word_display(word: str, *, template: str = SIGN_TEMPLATE, max_length: int = 100):
    if len(word) > max_length:
        raise errors.BadArgument("⚠️ Input too long. Try a shorter query.")
    quoted_word = quote_plus(word).lower()
    return template.format(
        word_uppercased=word.upper(),
        howsign=f"https://howsign.app/?s={quoted_word}",
        lifeprint=f"https://www.google.com/search?&q=site%3Alifeprint.com+{quoted_word}",
        handspeak=f"https://www.google.com/search?&q=site%3Ahandspeak.com+{quoted_word}",
        signingsavvy=f"https://www.signingsavvy.com/search/{quoted_word}",
        spread_the_sign=f"https://www.spreadthesign.com/en.us/search/?q={quoted_word}",
        youglish=f"https://youglish.com/pronounce/{quoted_word}/signlanguage/asl",
    )


def sign_impl(word: str):
    logger.info(f"sending links for: '{word}'")
    spoiler = get_spoiler_text(word)
    word = spoiler if spoiler else word
    has_multiple = "," in word
    template = SIGN_SPOILER_TEMPLATE if spoiler else SIGN_TEMPLATE
    if has_multiple:
        words = word.split(",")
        embed = discord.Embed()
        for word in words:
            word = word.strip()
            title = f"||{word.upper()}||" if spoiler else word.upper()
            embed.add_field(name=title, value=word_display(word, template=template))
    else:
        title = f"||{word.upper()}||" if spoiler else word.upper()
        embed = discord.Embed(
            title=title,
            description=word_display(word, template=template),
        )

    return {"embed": embed}


HANDSHAPE_HELP = """Show a random or specific handshape

Examples:
{COMMAND_PREFIX}handshape
{COMMAND_PREFIX}handshape claw5

Enter {COMMAND_PREFIX}handshapes to show a list of handshapes.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshape_impl(name: str):
    logger.info(f"handshape: '{name}'")
    try:
        if name == "random":
            handshape = handshapes.get_random_handshape()
        else:
            handshape = handshapes.get_handshape(name)
    except handshapes.HandshapeNotFoundError:
        logger.info(f"handshape '{name}' not found")
        suggestion = did_you_mean(name, tuple(handshapes.HANDSHAPES.keys()))
        if suggestion:
            return {
                "content": f'"{name}" not found. Did you mean "{suggestion}"? Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }
        else:
            return {
                "content": f'"{name}" not found. Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }

    filename = f"{handshape.name}.png"
    file_ = discord.File(handshape.path, filename=filename)
    embed = discord.Embed(title=handshape.name)
    embed.set_image(url=f"attachment://{filename}")
    return {
        "file": file_,
        "embed": embed,
    }


HANDSHAPES_TEMPLATE = """{handshapes}

Enter `{COMMAND_PREFIX}handshape` to display a random handshape or `{COMMAND_PREFIX}handshape [name]` to display a specific handshape.
"""

HANDSHAPES_HELP = """List handshapes

Enter {COMMAND_PREFIX}handshape to display a random handshape or {COMMAND_PREFIX}handshape [name] to display a specific handshape.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshapes_impl():
    return {
        "content": HANDSHAPES_TEMPLATE.format(
            handshapes=", ".join(handshapes.HANDSHAPES.keys()),
            COMMAND_PREFIX=COMMAND_PREFIX,
        )
    }


class ASL(Cog):
    def __init__(self, bot: Bot):
        self.bot = bot

    @command(name="sign", aliases=("howsign", COMMAND_PREFIX), help=SIGN_HELP)
    async def sign_command(self, ctx: Context, *, word: str):
        await ctx.send(**sign_impl(word))

    @sign_command.error
    async def sign_error(self, ctx: Context, error: Exception):
        # Ignore "??"
        if isinstance(error, errors.MissingRequiredArgument):
            if ctx.invoked_with == COMMAND_PREFIX:
                logger.info(
                    f"no argument passed to {COMMAND_PREFIX}{ctx.invoked_with}. ignoring..."
                )
            else:
                await ctx.send(
                    f"⚠️ Enter a word or phrase to search for after `{COMMAND_PREFIX}{ctx.invoked_with}`."
                )

    @command(name="handshape", aliases=("shape",), help=HANDSHAPE_HELP)
    async def handshape_command(self, ctx: Context, name="random"):
        await ctx.send(**handshape_impl(name))

    @command(name="handshapes", aliases=("shapes",), help=HANDSHAPES_HELP)
    async def handshapes_command(self, ctx: Context):
        logger.info("sending handshapes list")
        await ctx.send(**handshapes_impl())


def setup(bot: Bot) -> None:
    bot.add_cog(ASL(bot))
