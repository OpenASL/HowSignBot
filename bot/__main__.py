import logging

from aiohttp import web

from . import __version__, settings
from .app import app

log_format = "%(asctime)s - %(name)s %(levelname)s: %(message)s"
logging.getLogger("discord").setLevel(logging.WARNING)
logging.basicConfig(
    format=log_format,
    level=settings.LOG_LEVEL,
)
logger = logging.getLogger(__name__)

logger.info(f"starting bot version {__version__}")
web.run_app(app, port=settings.PORT)
