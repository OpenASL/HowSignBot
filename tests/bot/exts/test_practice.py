import os
import random
from unittest import mock

import gspread
import pytest
import pytz
from asynctest import patch
from disnake.ext import commands
from freezegun import freeze_time

# Must be before bot import
os.environ["TESTING"] = "true"

from bot import database, settings  # noqa:E402
from bot.exts.practices import practice  # noqa:E402

random.seed(1)


@pytest.fixture
async def mock_worksheet(monkeypatch, db):
    monkeypatch.setattr(settings, "GOOGLE_PRIVATE_KEY", "fake", raising=True)
    await db.execute(
        database.guild_settings.insert(), {"guild_id": 1234, "schedule_sheet_key": "abc"}
    )
    # Need to mock both of these since they get called in different files
    with patch(
        "bot.exts.practices.practice.get_practice_worksheet_for_guild"
    ) as mock_get_worksheet, patch(
        "bot.exts.practices._practice_sessions.get_practice_worksheet_for_guild"
    ) as mock_get_worksheet2:
        WorksheetMock = mock.Mock(spec=gspread.Worksheet)
        WorksheetMock.get_all_values.return_value = [
            ["docs", "more docs", ""],
            ["Start time", "Host (optional)", "Details (optional)", ""],
            ["Friday 5pm EDT", "Steve", "<@!12345>", "recurring", ""],
            ["Friday 6pm EDT", "Steve", "<@!12345>", "paused", "x"],
            ["Wed 6pm edt", "", "", "another recurring", ""],
            ["9/26 2pm PDT 2020", "Steve", "<@!12345>", "one time", ""],
            ["Sunday, September 27 02:00 PM EDT 2020", "", "", "another 1Ô∏è‚É£", ""],
            ["Nov 26 02:00 PM EDT 2020", "", "", "Turkey day PRACTICE", ""],
        ]
        mock_get_worksheet.return_value = mock_get_worksheet2.return_value = WorksheetMock
        yield WorksheetMock


@pytest.mark.parametrize(
    "when",
    (
        None,
        "today",
        "tomorrow",
        "saturday",
        "9/27",
    ),
)
@pytest.mark.asyncio
@freeze_time("2020-09-25 14:00:00")
async def test_schedule(snapshot, mock_worksheet, store, when):
    result = await practice.schedule_impl(1234, when)
    assert result == snapshot


@pytest.mark.asyncio
@freeze_time("2020-11-26 14:00:00")
async def test_schedule_on_a_holiday(snapshot, mock_worksheet, store):
    result = await practice.schedule_impl(1234, None)
    assert result == snapshot


@pytest.mark.asyncio
@freeze_time("2020-09-25 14:00:00")
async def test_schedule_no_practices(snapshot, mock_worksheet):
    result = await practice.schedule_impl(1234, "9/28/2020")
    embed = result["embed"]
    assert "September 28" in embed.description
    assert "There are no scheduled practices today" in embed.description
    assert (
        "To schedule a practice, edit the schedule below or use the" in embed.description
    )


@pytest.mark.parametrize(
    "start_time",
    (
        "2pm edt",
        "at 2pm edt",
        '2pm edt "chat and games! üéâ"',
        "9/24 1:45pm edt ‚Äúaround 45 min.-1 hour‚Äù",
        "tomorrow 2pm pdt",
        "sunday 10:30pm pdt",
        "9/27 9am cdt",
        '"games" 10am edt Sunday',
        '"games" at 10am edt on Sunday',
        '"games" at 10am edt on 9/27',
        '10am edt "games" on 9/27',
        '"classifiers" at 6pm pdt',
    ),
)
@freeze_time("2020-09-25 14:00:00")
@pytest.mark.asyncio
async def test_practice(snapshot, mock_worksheet, start_time, store):
    await practice.practice_impl(
        guild_id=1234,
        host="Steve",
        mention="<@!12345>",
        start_time=start_time,
        user_id=12345679,
    )
    mock_worksheet.append_row.assert_called_once()
    assert mock_worksheet.append_row.call_args[0][0] == snapshot


@pytest.mark.asyncio
async def test_practice_caches_timezone(mock_worksheet, store):
    await practice.practice_impl(
        guild_id=1234,
        host="Steve",
        mention="<@!12345>",
        start_time="tomorrow 8pm est",
        user_id=123,
    )
    timezone = await store.get_user_timezone(123)
    assert timezone == pytz.timezone("America/New_York")


@pytest.mark.asyncio
async def test_practice_errors_if_time_zone_cannot_be_parsed(mock_worksheet, store):
    with pytest.raises(commands.errors.BadArgument, match="Could not parse time zone"):
        await practice.practice_impl(
            guild_id=1234,
            host="Steve",
            mention="<@!12345>",
            start_time="tomorrow 8pm",
            user_id=321,
        )


@pytest.mark.asyncio
async def test_practice_nearby_date(snapshot, mock_worksheet, store):
    with freeze_time("2020-09-25 14:00:00"):
        await practice.practice_impl(
            guild_id=1234,
            host="Steve",
            mention="<@!12345>",
            start_time="10:30am edt",
            user_id=123,
        )

    mock_worksheet.append_row.assert_called_once()
    appended_row = mock_worksheet.append_row.call_args[0][0]
    assert appended_row == (
        "Friday, September 25 10:30 AM EDT 2020",
        "Steve",
        "<@!12345>",
        "",
    )


@pytest.mark.parametrize(
    "start_time",
    (
        "today",
        "tomorrow",
        "today edt",
        "today pdt",
    ),
)
@freeze_time("2020-09-25 14:00:00")
@pytest.mark.asyncio
async def test_practice_common_mistakes(snapshot, mock_worksheet, start_time):
    with pytest.raises(
        commands.errors.CommandError, match="Ô∏èTo schedule a practice, enter a time"
    ):
        await practice.practice_impl(
            guild_id=1234,
            host="Steve",
            mention="<@!12345>",
            start_time=start_time,
            user_id=123,
        )
    mock_worksheet.append_row.assert_not_called()


@freeze_time("2020-09-25 14:00:00")
@pytest.mark.asyncio
async def test_practice_invalid(mock_worksheet, store):
    with pytest.raises(
        commands.errors.CommandError,
        match='Could not parse "invalid" into a date or time. Make sure to include "am" or "pm" as well as a timezone',
    ):
        await practice.practice_impl(
            guild_id=1234,
            host="Steve",
            mention="<@!12345>",
            start_time="invalid",
            user_id=123,
        )
    mock_worksheet.append_row.assert_not_called()
