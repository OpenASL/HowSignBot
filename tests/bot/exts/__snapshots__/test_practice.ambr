# name: test_practice["classifiers" at 6pm pdt]
  <class 'tuple'> (
    'Friday, September 25 06:00 PM PDT 2020',
    'Steve',
    '<@!12345>',
    'classifiers',
  )
---
# name: test_practice["games" 10am edt Sunday]
  <class 'tuple'> (
    'Sunday, September 27 10:00 AM EDT 2020',
    'Steve',
    '<@!12345>',
    'games',
  )
---
# name: test_practice["games" at 10am edt on 9/27]
  <class 'tuple'> (
    'Sunday, September 27 10:00 AM EDT 2020',
    'Steve',
    '<@!12345>',
    'games',
  )
---
# name: test_practice["games" at 10am edt on Sunday]
  <class 'tuple'> (
    'Sunday, September 27 10:00 AM EDT 2020',
    'Steve',
    '<@!12345>',
    'games',
  )
---
# name: test_practice[10am edt "games" on 9/27]
  <class 'tuple'> (
    'Sunday, September 27 10:00 AM EDT 2020',
    'Steve',
    '<@!12345>',
    'games',
  )
---
# name: test_practice[2pm edt "chat and games! \U0001f389"]
  <class 'tuple'> (
    'Friday, September 25 02:00 PM EDT 2020',
    'Steve',
    '<@!12345>',
    'chat and games! ðŸŽ‰',
  )
---
# name: test_practice[2pm edt]
  <class 'tuple'> (
    'Friday, September 25 02:00 PM EDT 2020',
    'Steve',
    '<@!12345>',
    '',
  )
---
# name: test_practice[9/24 1:45pm edt \u201caround 45 min.-1 hour\u201d]
  <class 'tuple'> (
    'Friday, September 24 01:45 PM EDT 2021',
    'Steve',
    '<@!12345>',
    'around 45 min.-1 hour',
  )
---
# name: test_practice[9/27 9am cdt]
  <class 'tuple'> (
    'Sunday, September 27 09:00 AM CDT 2020',
    'Steve',
    '<@!12345>',
    '',
  )
---
# name: test_practice[at 2pm edt]
  <class 'tuple'> (
    'Friday, September 25 02:00 PM EDT 2020',
    'Steve',
    '<@!12345>',
    '',
  )
---
# name: test_practice[sunday 10:30pm pdt]
  <class 'tuple'> (
    'Sunday, September 27 10:30 PM PDT 2020',
    'Steve',
    '<@!12345>',
    '',
  )
---
# name: test_practice[tomorrow 2pm pdt]
  <class 'tuple'> (
    'Saturday, September 26 02:00 PM PDT 2020',
    'Steve',
    '<@!12345>',
    '',
  )
---
# name: test_schedule[None]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=<Colour value=15105570>,
      colour=<Colour value=15105570>,
      description='Today - Friday, September 25',
      fields=<class 'list'> [
        EmbedProxy(inline=False, name='2 PM PDT / 3 PM MDT / 4 PM CDT / 5 PM EDT', value='[+Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=ASL+Practice%3A+recurring&dates=20200925T210000Z%2F20200925T220000Z&details=See+the+full+schedule+here%3A+https%3A%2F%2Fdocs.google.com%2Fspreadsheets%2Fd%2Fabc%2Fedit)\n> Host: <@!12345>\n> Details: recurring'),
        EmbedProxy(inline=True, name='ðŸ—“ View or edit the schedule using the link below.', value='[Full schedule](https://docs.google.com/spreadsheets/d/abc/edit)'),
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title=Embed.Empty,
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
# name: test_schedule[today]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=<Colour value=15105570>,
      colour=<Colour value=15105570>,
      description='Today - Friday, September 25',
      fields=<class 'list'> [
        EmbedProxy(inline=False, name='2 PM PDT / 3 PM MDT / 4 PM CDT / 5 PM EDT', value='[+Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=ASL+Practice%3A+recurring&dates=20200925T210000Z%2F20200925T220000Z&details=See+the+full+schedule+here%3A+https%3A%2F%2Fdocs.google.com%2Fspreadsheets%2Fd%2Fabc%2Fedit)\n> Host: <@!12345>\n> Details: recurring'),
        EmbedProxy(inline=True, name='ðŸ—“ View or edit the schedule using the link below.', value='[Full schedule](https://docs.google.com/spreadsheets/d/abc/edit)'),
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title=Embed.Empty,
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
# name: test_schedule_on_a_holiday
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=<Colour value=15105570>,
      colour=<Colour value=15105570>,
      description='Today - Thursday, November 26 ðŸ¦ƒ',
      fields=<class 'list'> [
        EmbedProxy(inline=False, name='11 AM PST / 12 PM MST / 1 PM CST / 2 PM EST', value='[+Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=ASL+Practice%3A+Turkey+day+PRACTICE&dates=20201126T190000Z%2F20201126T200000Z&details=See+the+full+schedule+here%3A+https%3A%2F%2Fdocs.google.com%2Fspreadsheets%2Fd%2Fabc%2Fedit)\n> Details: Turkey day PRACTICE'),
        EmbedProxy(inline=True, name='ðŸ—“ View or edit the schedule using the link below.', value='[Full schedule](https://docs.google.com/spreadsheets/d/abc/edit)'),
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title=Embed.Empty,
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
